src/main/java/gr8pefish/ironbackpacks/util/registry/UpgradeRegistry.java

```diff
@@ -0,0 +1,172 @@
+     * sane defaults for avoiding boilerplate. See methods and their javadocs for
+     * details on customizable properties.
+     *
+     * @author EngiN33R
+     * @since 4.0.0
+     */
+    public static class Builder {
+        private static final Item.Settings SETTINGS = new Item.Settings().group(ItemGroup.MISC);
+
+        private Item.Settings settings = SETTINGS;
+        private int points = 1;
+        private int tier = 0;
+
+        /**
+         * Sets the upgrade point cost for this upgrade. Default is 1.
+         *
+         * @param points upgrade point cost
+         * @return builder instance for chaining
+         * @since 4.0.0
+         */
+        public Builder points(@Nonnegative int points) {
+            this.points = points;
+            return this;
+        }
+
+        /**
+         * Sets the minimum tier of backpack for this upgrade. Default is 0.
+         *
+         * @param tier minimum backpack tier
+         * @return builder instance for chaining
+         * @since 4.0.0
+         */
+        public Builder tier(@Nonnegative int tier) {
+            this.tier = tier;
+            return this;
+        }
+
+        /**
+         * Sets the item settings to pass to the item on creation. Defaults to
+         * being stackable and categorized under Miscellaneous in the creative menu.
+         *
+         * @param settings item settings
+         * @return builder instance for chaining
+         * @since 4.0.0
+         */
+        public Builder settings(Item.Settings settings) {
+            this.settings = settings;
+            return this;
+        }
+
+        private BaseUpgradeItem build() {
+            return new BaseUpgradeItem(settings, points, tier);
+        }
+
+        /**
+         * Registers the upgrade specification as an actual item. The item ID
+         * will match the upgrade ID. This is the terminal operation in a builder chain.
+         *
+         * @param id identifier to register the item under
+         * @since 4.0.0
+         */
+        public void register(@Nonnull Identifier id) {
+            UpgradeRegistry.register(id, build());
+        }
+    }
+
+    /**
+     * Instantiates a builder for an upgrade.
+     *
+     * @see Builder
+     * @return builder instance
+     * @since 4.0.0
+     */
+    public static Builder create() {
+        return new Builder();
+    }
+
+    /**
+     * Registers the provided upgrade item instance in the upgrade registry and
+     * as an item.
+     *
+     * @param id upgrade and item ID
+     * @param item item instance
+     */
+    public static void register(@Nonnull Identifier id, @Nonnull UpgradeItem item) {
+        if (id.equals(new Identifier(MODID, "upgrade/blank"))) {
+            LOGGER.warn("Trying to overwrite blank upgrade! Skipping...");
+            return;
+        }
+        if (UPGRADES.putIfAbsent(id, item) != null) {
+            LOGGER.warn("Upgrade with ID " + id.toString() + " already exists! Skipping...");
```

Comment:
Logger has it's own form of string formatting which should be used.
```suggestion
            LOGGER.warn("Upgrade with ID {} already exists! Skipping...", id);
```